### PR TITLE
add basic contextual commands

### LINKED ISSUE #151937: Explore contextual actions in the terminal
Part of this will be building the foundation of listening to commands/args that would also get used in https://github.com/microsoft/vscode/issues/145234

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #151937

### CODE REVIEW

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (meganrogge) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:12]: perhaps this should be deferred to the terminal instance as well

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (meganrogge) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:238]: to do: better naming of variables - defer to helper function bc it's massive

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (meganrogge) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:293]: to do: figure out styling to put decoration on top of command decoration and dispose of it when the context menu is hidden or add `don't show again` action

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:208]: revert

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (Tyriar) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:280]: The decoration stuff would be better to live outside of the capability (this might have been like this in my rough exploration)

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (Tyriar) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:308]: This sort of thing is fine on CommandDetectionCapability but we should be more thoughtful on an ITerminalInstance API. Something like this:

```ts
interface ITerminalInstance {
  registerContextualAction(options: ITerminalContextualActionOptions): IDisposable;
}

interface ITerminalContextualActionOptions {
  commandLineMatcher: string | RegExp;
  // Allows searching a subset of lines efficiently
  outputMatcher?: ITerminalOutputMatcher;
  // Allow the name to be generated based on capture groups in the output matcher(s)
  actionName: string | DynamicActionName;
  callback: ContextualActionCallback;
}

type DynamicActionName = (commandLineMatch: RegExpMatchArray, outputMatch?: RegExpMatchArray | null) => string;
type ContextualActionCallback = (commandLineMatch: RegExpMatchArray, outputMatch?: RegExpMatchArray | null) => void;

interface ITerminalOutputMatcher {
	lineMatcher: string | RegExp; 
	anchor?: 'top' | 'bottom'
	offset?: number;
	length?: number;
}
```

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (Tyriar) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:12]: They should get registered into the terminal instance, we may want the main list that applies to all instances to live in terminal service or terminal instance service so they can be initialized when the instance is created?

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (Tyriar) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:308]: If needed we could also have command and arg matchers so that consumers don't need to each parse the command line if they only apply on a certain argument.

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts
Comment (Tyriar) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:12]: Alternatively require any consumers to register them into each newly created terminal instance.

Comment (meganrogge) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts]: wouldn't it make more sense to keep the registration on the `contextActionAddon` which is accessible via the terminal instance by way of `xterm.contextAction`?

File: src/vs/platform/terminal/node/terminalProcess.ts
Comment (meganrogge) [inline src/vs/platform/terminal/node/terminalProcess.ts:198]: revert

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:214]: having issues w this and not sure why

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:214]: also had to change for the test bc the test was failing when instantiation service was used

**Comment by meganrogge** [comment]
to do:

- additional regexes for port in use
- tests?

File: src/vs/platform/terminal/common/terminal.ts
Comment (Tyriar) [inline src/vs/platform/terminal/common/terminal.ts:656]: Don't bother with that generic child process refactor I was talking about, we'll probably move all the core ones into a builtin extension eventually.

File: src/vs/platform/terminal/node/ptyService.ts
Comment (Tyriar) [inline src/vs/platform/terminal/node/ptyService.ts:109]: This will need a separate one for Windows

File: src/vs/workbench/contrib/terminal/browser/terminal.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:462]: `contextualActions`?

File: src/vs/workbench/contrib/terminal/browser/terminalActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalActions.ts:373]: Yeah a separate method is probably best so this sort of thing could be avoided.

File: src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts:14]: This is a bit hard to read, I think `\d\.\d\.\d\.\d` would be easier to understand for an IP. Also I don't understand the \s here: `(\d\s*){3}\d`

File: src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts:42]: I was thinking since running a command is one of the most common things we should elevate the concept into the contextual action options. This will let us make the formatting of each quick fix consistently controlled by a single spot. Example:

```ts
getActions(matchResult: ContextualMatchResult, command: ITerminalCommand): IAction | IRunCommand;

interface IRunCommand {
  command: string;
}
```

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:146]: We don't want to expose this, is it just used in tests? If so _processManager could be protected and exposed in a sub class.

File: src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts:105]: This in and any shouldn't be needed?

**Review by Tyriar — DISMISSED** [review]

File: src/vs/platform/terminal/common/terminal.ts
Comment (meganrogge) [reply src/vs/platform/terminal/common/terminal.ts:656]: yes agree, that's what I thought too

File: src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: I noticed that the line can wrap mid stream like

```
https://1.2.3.4:0
555
```

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: maybe that's only because of how i've written the test though?

File: src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts:105]: it is saying `Property 'registerDecoration' does not exist on type 'Terminal'.ts(2339)` when I remove that

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts:105]: oh that was bc I had imported `Terminal` from `xterm-headless`

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:146]: hm tried that and not sure how to get that to work, so exposed the function on terminal instance

File: src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: yeah I was thinking that too. thanks for the reminder. we will prob want more properties on `IRunCommand` so we don't have to guess though

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: probably sufficient to have `IRunCommand` extend `IAction`

**Comment by meganrogge** [comment]
@Tyriar I'd like to defer the windows work and merge this today if you're still around

File: src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: for example, didn't work here bc
`a
ddress`
Screen Shot 2022-09-21 at 9 46 27 AM

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: so going to [].join('\s+') between all to avoid all of these possibilities

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: oh this is quite complicated bc really need to join each of the chars in a word with `\s*`

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts]: added tests and improved regex

**Comment by meganrogge** [comment]
Just need to test on windows

**Comment by meganrogge** [comment]
doesn't currently work for windows bc command finished doesn't happen
tasks

**Review by Tyriar — APPROVED** [review]

**Pull request merged**
